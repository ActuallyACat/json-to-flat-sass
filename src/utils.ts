#!/usr/bin/env node
import * as fs from 'fs';
import * as path from 'path';
import {promises} from 'dns';

// Inspired by: https://blog.grossman.io/how-to-write-async-await-without-try-catch-blocks-in-javascript/
export default function to(promise) {
  return promise
    .then(data => {
      return [null, data];
    })
    .catch(err => [err]);
}

const readFilePromise = (filename: string) => {
  return new Promise(function(resolve, reject) {
    fs.readFile(filename, (err, data) => {
      if (err) reject(err);
      resolve(data);
    });
  });
};

const writeFilePromise = (filename: string, content: string) => {
  return new Promise((resolve, reject) => {
    fs.writeFile(filename, content, err => {
      if (err) reject(err);
      resolve();
    });
  });
};

export const jsonToFlatSass = async ({
  source,
  destination,
  separator,
  variablePrefix = '$'
}: {
  source: string;
  destination: string;
  separator: string;
  variablePrefix?: '$' | '@'
}) => {
  let flattenedInput = '',
    err: null | Error,
    fileData: null;

  [err, fileData] = await to(readFilePromise(source));
  if (!fileData) {
    throw new Error('Error reading source file');
  }

  let parsedString = {};
  try {
    parsedString = JSON.parse(fileData);
  } catch (err) {
    throw new Error('Error parsing input as JSON');
  }

  flattenedInput = flattenInputToString(parsedString, separator, variablePrefix);
  const {dir} = path.parse(destination);
  await fs.mkdir(dir, {recursive: true}, () => {
    // an error here means dir is a path. Fine to squash
  });
  [err] = await to(writeFilePromise(destination, flattenedInput));
  if (err) {
    throw new Error('Error writing to destination file');
  }
};

export const flattenInputToString = (
  jsonSource: object,
  separator: string,
  variablePrefix: string
): string => {
  let flatMap = `/** THIS IS AN AUTOGENERATED FILE **/\n\n`;

  const dive = (input: {[index: string]: any}, name: string): void => {
    if (typeof input !== 'object') {
      flatMap += `${variablePrefix}${name}: ${input};\n`;
      return;
    }
    Object.keys(input).forEach(key => {
      let concatName = name === '' ? key : `${name}${separator}${key}`;
      return dive(input[key], concatName);
    });
  };

  if (typeof jsonSource === 'object') {
    dive(jsonSource, '');
  }

  return flatMap;
};
