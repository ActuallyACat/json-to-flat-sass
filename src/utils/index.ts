const fsPromises = require('fs').promises; // using experimental API -> https://github.com/nodejs/node/issues/21014

export const jsonToFlatSass = async ({
  source,
  destination,
  separator = '-',
}: {
  source: string;
  destination?: string;
  separator?: string;
}) => {
  let flattenedInput = '';
  try {
    const sourceFile = await fsPromises.readFile(source, 'utf8');
    const parsedString = JSON.parse(sourceFile);
    flattenedInput = flattenInputToString(parsedString, separator);
  } catch (err) {
    console.error(err);
    return;
  }

  if (destination !== undefined) {
    await fsPromises.writeFile(destination, flattenedInput);
  } else {
    console.log(flattenedInput);
  }
};

export const flattenInputToString = (
  jsonSource: object,
  separator: string,
): string => {
  let flatMap = '';

  const dive = (input: {[index: string]: any}, name: string): void => {
    if (typeof input !== 'object') {
      flatMap += `$${name} = '${input}';\n`;
      return;
    }
    Object.keys(input).forEach(key => {
      let concatName = name === '' ? key : `${name}${separator}${key}`;
      return dive(input[key], concatName);
    });
  };

  if (typeof jsonSource === 'object') {
    dive(jsonSource, '');
  }

  return flatMap;
};
