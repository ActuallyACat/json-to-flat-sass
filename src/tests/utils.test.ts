import {flattenInputToString} from '../utils';
import mocks from './mocks';

describe('flattenInput', () => {
  it('should flatten and insert correct separator for sass', () => {
    mocks.forEach(json => {
      expect(flattenInputToString(json, '-', '$')).toMatchSnapshot();
    });
  });

  it('should flatten and insert correct separator for less', () => {
    mocks.forEach(json => {
      expect(flattenInputToString(json, '-', '@')).toMatchSnapshot();
    });
  });
});
