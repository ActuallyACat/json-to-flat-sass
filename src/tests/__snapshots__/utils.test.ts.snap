// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flattenInput should flatten and insert correct separator for less 1`] = `
"/** THIS IS AN AUTOGENERATED FILE **/

@purple-100: #F0F1F4;
@purple-200: #E1E2EA;
@purple-300: #C4C5D4;
@purple-400: #A7A9C1;
@purple-500: #898BA9;
@purple-600: #5B5E7E;
@purple-700: #4B4D68;
@purple-800: #35374A;
@blue-100: #E6F0F7;
@blue-200: #CCE1F0;
@blue-300: #99C3E1;
@blue-400: #4D95CA;
@blue-500: #0168B3;
@green-100: #E9F6F3;
@green-200: #BCE3DC;
@green-300: #90D1C5;
@green-400: #64BFAE;
@green-500: #21A38B;
@green-600: #177261;
@yellow-100: #FFFDF4;
@yellow-200: #FFF5D2;
@yellow-300: #FFEBA5;
@yellow-400: #FFE278;
@yellow-500: #FFCE1E;
@red-100: #FDEDEF;
@red-200: #FBC9CE;
@red-300: #F6949E;
@red-400: #F3707D;
@red-500: #F04C5D;
@red-600: #A83541;
@orange-100: #FFF1EE;
@orange-200: #FDD6CD;
@orange-300: #FDBAAB;
@orange-400: #FC9E8A;
@orange-500: #FA7558;
"
`;

exports[`flattenInput should flatten and insert correct separator for less 2`] = `
"/** THIS IS AN AUTOGENERATED FILE **/

@heading: \\"Greycliff CF\\";
@paragraph: \\"Inter\\";
@preformatted: \\"IBM Plex Mono\\";
"
`;

exports[`flattenInput should flatten and insert correct separator for sass 1`] = `
"/** THIS IS AN AUTOGENERATED FILE **/

$purple-100: #F0F1F4;
$purple-200: #E1E2EA;
$purple-300: #C4C5D4;
$purple-400: #A7A9C1;
$purple-500: #898BA9;
$purple-600: #5B5E7E;
$purple-700: #4B4D68;
$purple-800: #35374A;
$blue-100: #E6F0F7;
$blue-200: #CCE1F0;
$blue-300: #99C3E1;
$blue-400: #4D95CA;
$blue-500: #0168B3;
$green-100: #E9F6F3;
$green-200: #BCE3DC;
$green-300: #90D1C5;
$green-400: #64BFAE;
$green-500: #21A38B;
$green-600: #177261;
$yellow-100: #FFFDF4;
$yellow-200: #FFF5D2;
$yellow-300: #FFEBA5;
$yellow-400: #FFE278;
$yellow-500: #FFCE1E;
$red-100: #FDEDEF;
$red-200: #FBC9CE;
$red-300: #F6949E;
$red-400: #F3707D;
$red-500: #F04C5D;
$red-600: #A83541;
$orange-100: #FFF1EE;
$orange-200: #FDD6CD;
$orange-300: #FDBAAB;
$orange-400: #FC9E8A;
$orange-500: #FA7558;
"
`;

exports[`flattenInput should flatten and insert correct separator for sass 2`] = `
"/** THIS IS AN AUTOGENERATED FILE **/

$heading: \\"Greycliff CF\\";
$paragraph: \\"Inter\\";
$preformatted: \\"IBM Plex Mono\\";
"
`;
